File: built-in plugins dsp/mb-fm-bass.cpp
/**
 * MB FM Bass
 * Category : instrument
 * Type     : fm
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Punchy FM bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FM_BASS_H
#define MB_FM_BASS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFmBass : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-fm-bass";
    static constexpr const char* PLUGIN_NAME    = "MB FM Bass";
    static constexpr const char* PLUGIN_TYPE    = "fm";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float modIndex = 4f;  // range [0, 15]
    float punch = 0.7f;  // range [0, 1]
    float volume = 0.85f;  // range [0, 1]
    };

    MbFmBass() = default;
    ~MbFmBass() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.modIndex = std::clamp(params.modIndex, 0f, 15f);
        params.punch = std::clamp(params.punch, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB FM Bass
        return input;
    }
};

#endif // MB_FM_BASS_H

File: built-in plugins dsp/mb-fm-bass.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFmBassPlugin: PluginDefinition = { id: 'mb-fm-bass', slug: 'mb-fm-bass', name: 'MB FM Bass', category: 'instrument', type: 'fm', version: '1.0.0', description: 'Punchy FM bass', author: 'Max Booster', oscillators: [{ type: 'sine', detune: 0, gain: 0.9 }], envelope: { attack: 0.001, decay: 0.2, sustain: 0.6, release: 0.15 }, parameters: [{ id: 'modIndex', name: 'Mod Index', type: 'float', defaultValue: 4, minValue: 0, maxValue: 15, automatable: true }, { id: 'punch', name: 'Punch', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.85, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { modIndex: 4, punch: 0.7, volume: 0.85 } };

export default MbFmBassPlugin;
